sarc/cache.py

```diff
@@ -64,6 +67,208 @@ class CacheException(Exception):
+        from the given date and continuing forward through all subsequent dates.
+
+        Args:
+            from_time: The earliest datetime to include in results.
+
+        Yields:
+            Path: The path for each matching cache entry.
+        """
+        cdir = self.cache_dir
+        from_time = ensure_utc(from_time)
+
+        first_dir = self._dir_from_date(cdir, from_time)
+
+        if first_dir.exists():
+            from_time_nodays = from_time.time()
+            for file in filter(
+                lambda fname: time.fromisoformat(fname.parts[-1]) >= from_time_nodays,
+                sorted(first_dir.iterdir()),
+            ):
+                yield file
+
+        from_time = from_time.replace(hour=0, minute=0, second=0, microsecond=0)
+        from_time += timedelta(days=1)
+
+        first_year_done = False
+        first_month_done = False
+
+        for year_dir in sorted(
+            filter(lambda y: int(y.parts[-1]) >= from_time.year, cdir.iterdir())
+        ):
+            if not first_year_done and int(year_dir.parts[-1]) > from_time.year:
+                first_year_done = True
+            for month_dir in sorted(
+                filter(
+                    lambda m: first_year_done or int(m.parts[-1]) >= from_time.month,
+                    year_dir.iterdir(),
+                )
+            ):
+                if not first_month_done and int(month_dir.parts[-1]) > from_time.month:
+                    first_month_done = True
+                for day_dir in sorted(
+                    filter(
+                        lambda d: first_month_done or int(d.parts[-1]) >= from_time.day,
+                        month_dir.iterdir(),
+                    )
+                ):
+                    for file in sorted(day_dir.iterdir()):
+                        yield file
+                first_month_done = True
+            first_year_done = True
+
+    def read_from(self, from_time: datetime) -> Iterable[CacheEntry]:
+        """Read all cached entries starting from a specific datetime.
+
+        Returns an iterator over all cached entries that were created at or after
+        the specified time. Unlike `read_from()`, this method returns entries for
+        all keys, not just a specific key. The cache files are searched through
+        the date hierarchy starting from the given date and continuing forward
+        through all subsequent dates.
+
+        Args:
+            from_time: The earliest datetime to include in results. Must be UTC.
+
+        Yields:
+            tuple[str, bytes]: A tuple containing:
+                - The cache key
+                - The binary data from the cache entry
+
+        Example:
+            >>> cache = Cache("my_data")
+            >>> start_time = datetime(2024, 1, 15, 10, 0, 0)
+            >>> for key, data in cache.read_from_all(start_time):
+            ...     print(f"Key: {key}, Data size: {len(data)} bytes")
```

Comment:
Current method iterates over `CacheEntry` objects, not `key, data` pairs in CacheEntry objects.